core-common/src/main/java/org/zalando/nakadi/domain/kpi/KPIEvent.java

```diff
@@ -0,0 +1,65 @@
+package org.zalando.nakadi.domain.kpi;
+
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.generic.GenericRecordBuilder;
+import org.json.JSONObject;
+import org.zalando.nakadi.exceptions.runtime.NakadiBaseException;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.reflect.Field;
+import java.util.Arrays;
+import java.util.stream.Stream;
+
+public abstract class KPIEvent {
+
+    @Retention(RetentionPolicy.RUNTIME)
+    public @interface KPIField {
+        String value();
+    }
+
+    public static class KPIEventMappingException extends NakadiBaseException {
+        public KPIEventMappingException(final String message, final Throwable cause) {
+            super(message, cause);
+        }
+    }
+
+    /*
+        Implementation of this method should return the event-type name of the specific KPI event.
+        This method should not be confused with the getter "getEventType()" in some specific KPI events;
+        which is why this method is named "eventTypeOfThisKPIEvent()
+     */
+    public abstract String eventTypeOfThisKPIEvent();
+
+    private Stream<Field> getKPIFields() {
+        return Arrays.stream(this.getClass().getDeclaredFields())
```

Comment:
that works only with a higher level classes, what if I have more sophisticated structures ?

Reply:
Yes, the current implementation is only taking care of the simple structure with one level. 
I'm hoping to take care of the more complex structure once I do have a example for it.